07.ls_object/lib/ls_command.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+require 'optparse'
+require 'pathname'
+require_relative 'ls_file'
+
+class LsCommand
+  COLS = 3
+
+  def show_ls_files
+    options = parse_options
+    ls_files = build_ls_files(options)
+    if options[:long]
+      show_ls_files_in_long_format(ls_files)
+    else
+      show_ls_files_in_default_format(ls_files)
+    end
+  end
+
+  private
+
+  def parse_options
+    options = { all: false, reverse: false, long: false }
+
+    OptionParser.new do |opts|
+      opts.on('-a') { options[:all] = true }
+      opts.on('-r') { options[:reverse] = true }
+      opts.on('-l') { options[:long] = true }
+    end.parse!
+
+    options
+  end
+
+  def build_ls_files(options)
+    target_directory = Pathname(ARGV[0] || '.')
+    file_basenames = Dir.foreach(target_directory)
+    ls_files = file_basenames.map { |file_basename| LsFile.new(file_basename, target_directory) }
+    filtered_ls_files = options[:all] ? ls_files : ls_files.reject(&:hidden?)
+    options[:reverse] ? filtered_ls_files.sort.reverse : filtered_ls_files.sort
+  end
+
+  def show_ls_files_in_long_format(ls_files)
+    puts "total #{ls_files.sum(&:block_size)}"
+    max_length = find_ls_file_max_lengths(ls_files)
+    ls_files.each do |ls_file|
+      ls_file_properties = []
+      ls_file_properties << ls_file.mode
+      ls_file_properties << ls_file.hard_links.to_s.rjust(max_length[:hard_links])
+      ls_file_properties << ls_file.owner_name.to_s.ljust(max_length[:owner_name])
+      ls_file_properties << ls_file.owner_group_name.to_s.ljust(max_length[:owner_group_name])
+      ls_file_properties << ls_file.size_or_device_info.to_s.rjust(max_length[:size])
+      ls_file_properties << format_last_modified_time(ls_file)
+      ls_file_properties << format_name(ls_file)
+      puts ls_file_properties.join(' ')
+    end
+  end
+
+  def find_ls_file_max_lengths(ls_files)
+    {
+      hard_links: ls_files.map { |ls_file| ls_file.hard_links.to_s.length }.max,
+      owner_name: ls_files.map { |ls_file| ls_file.owner_name.to_s.length }.max,
+      owner_group_name: ls_files.map { |ls_file| ls_file.owner_group_name.to_s.length }.max,
+      size: ls_files.map { |ls_file| ls_file.size_or_device_info.to_s.length }.max
+    }
+  end
+
+  def format_last_modified_time(ls_file)
+    ls_file_last_modified_time = ls_file.last_modified_time
+    if Time.now.year == ls_file_last_modified_time.year
+      ls_file_last_modified_time.strftime('%b %e %H:%M')
+    else
+      ls_file_last_modified_time.strftime('%b %e  %Y')
+    end
+  end
+
+  def format_name(ls_file)
+    ls_file_path = ls_file.file_path
+    ls_file_name = ls_file.name
+    if File.symlink?(ls_file_path)
+      "#{ls_file_name} -> #{File.readlink(ls_file_path)}"
```

Comment:
```suggestion
    if ls_file.symlink?
      "#{ls_file.name} -> #{ls_file.referenced_file}"
```

このあたりもLsFile自身にメソッドを持たせるとよさそう

Reply:
`LsFile`に`symlink?`と`referenced_file`を新しく定義しました。